src/main/java/com/iabtcf/FieldDefs.java

```diff
@@ -0,0 +1,427 @@
+        public Integer apply(ByteBitVector t) {
+            return Integer.valueOf(t.readBits6(PPTC_NUM_CUSTOM_PURPOSES.getOffset(t)));
+        }
+
+        @Override
+        public boolean isDynamic() {
+            return true;
+        }
+    }),
+    PPTC_CUSTOM_PURPOSES_LI_TRANSPARENCY(new LengthSupplier() {
+        @Override
+        public Integer apply(ByteBitVector t) {
+            // same length as PPTC_CUSTOM_PURPOSES_CONSENT
+            return PPTC_CUSTOM_PURPOSES_CONSENT.getLength(t);
+        }
+
+        @Override
+        public boolean isDynamic() {
+            return true;
+        }
+    }),
+
+    // range entry, only field lengths are supported
+    NUM_ENTRIES(12, OffsetSupplier.NOT_SUPPORTED),
+    IS_A_RANGE(1, OffsetSupplier.NOT_SUPPORTED),
+    START_OR_ONLY_VENDOR_ID(16, OffsetSupplier.NOT_SUPPORTED),
+    END_VENDOR_ID(16, OffsetSupplier.NOT_SUPPORTED),
+
+    // publish restriction fields, only field lengths are supported
+    PURPOSE_ID(6, OffsetSupplier.NOT_SUPPORTED),
+    RESTRICTION_TYPE(2, OffsetSupplier.NOT_SUPPORTED),
+
+    // v1 fields
+    V1_VERSION(6, 0),
+    V1_CREATED(36),
+    V1_LAST_UPDATED(36),
+    V1_CMP_ID(12),
+    V1_CMP_VERSION(12),
+    V1_CONSENT_SCREEN(6),
+    V1_CONSENT_LANGUAGE(12),
+    V1_VENDOR_LIST_VERSION(12),
+    V1_PURPOSES_ALLOW(24),
+    V1_VENDOR_MAX_VENDOR_ID(16),
+    V1_VENDOR_IS_RANGE_ENCODING(1),
+    V1_VENDOR_DEFAULT_CONSENT(1),
+    V1_VENDOR_BITRANGE_FIELD(BitRangeFieldUtils.lengthSupplierV1());
+
+    private OffsetSupplier offset;
+    private LengthSupplier length;
+    private boolean isDynamic = false;
+    private boolean isDynamicInit = false;
+
+    FieldDefs(int length, FieldDefs field) {
+        assert field != this;
+
+        this.length = LengthSupplier.constant(length);
+        this.offset = OffsetSupplier.from(field);
+    }
+
+    FieldDefs(int length, OffsetSupplier offset) {
+        this.length = LengthSupplier.constant(length);
+        this.offset = offset;
+    }
+
+    FieldDefs(final int length, int offset) {
+        this.length = LengthSupplier.constant(length);
+        this.offset = OffsetSupplier.constant(offset);
+    }
+
+    FieldDefs(LengthSupplier length) {
+        this.length = length;
+        this.offset = OffsetSupplier.fromPrevious(this);
+    }
+
+    FieldDefs(final int length) {
+        this.length = LengthSupplier.constant(length);
+        this.offset = OffsetSupplier.fromPrevious(this);
+    }
+
+    /**
+     * Determines whether the field is dynamic, that is, it depends on the particular consent string
+     * being processed.
+     */
+    protected boolean isDynamic() {
+        if (!isDynamicInit) {
+            isDynamic = offset.isDynamic() || length.isDynamic();
+            isDynamicInit = true;
+        }
+
+        return isDynamic;
+    }
+
+    /**
+     * Returns the length of the field.
+     */
+    public int getLength(ByteBitVector bbv) {
+        return bbv.cache.getLength(this, length);
```

Comment:
Why not move getLength to ByteBitVector and have ByteBitVector internally call to the LengthOffsetCache. FieldDefs don't need to know that the length was cached somewhere, it is an implementation choice of the ByteBitVector.

Reply:
I'm not excited about having the cache be in the ByteBitVector let alone having it compute the lengths and offsets.